org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/JavaSearchTests.java

```diff
@@ -4593,4 +4576,108 @@ public void testAnonymousTypeMethodReferenceJarSearchGh375() throws Exception {
 		deleteProject("P");
 	}
 }
+/*
+ * Test that we can find method references called in an anonymous type defined in a method.
+ * https://github.com/eclipse-jdt/eclipse.jdt.core/issues/432
+ */
+public void testAnonymousTypeMethodReferenceSearchGh432() throws Exception {
+	String snippet1 = "package p;\n" +
+			"public class TestGh432 {\n" +
+			"  public static void main(String[] args) {\n" +
+			"    class R implements Runnable {\n" +
+			"      public void run() {\n" +
+			"        java.util.stream.Stream.empty().filter(TestGh432::missingReference).count();\n" +
+			"      }\n" +
+			"    }\n" +
+			"    new R().run();\n" +
+			"  }\n" +
+			"  public static boolean missingReference(Object arg) {\n" +
+			"    return true;\n" +
+			"  }\n" +
+			"}";
+	String snippet2 = "package p;\n" +
+			"public class TestGh432 {\n" +
+			"  private static void foo(Runnable r) {\n" +
+			"    r.run();\n" +
+			"  }\n" +
+			"  public static void main(String[] args) {\n" +
+			"    foo(new Runnable() {\n" +
+			"      public void run() {\n" +
+			"        java.util.stream.Stream.empty().filter(TestGh432::missingReference).count();\n" +
+			"      }\n" +
+			"    });\n" +
+			"  }\n" +
+			"  public static boolean missingReference(Object arg) {\n" +
+			"    return true;\n" +
+			"  }\n" +
+			"}";
+	String snippet3 = "package p;\n" +
```

Comment:
Any idea why the test would fail for this case?
I see that the method body of main is retained but still the class defined in the lambda is not processed such the method reference is not found.

```java
class TestGh432 {
	private static void foo(Runnable r) {
		r.run();
	}

	public static void main(String[] args) {
		foo(() -> {
			new Runnable() {
				public void run() {
					java.util.stream.Stream.generate(TestGh432::missingReference).count();
				}
			};
		});
	}

	public static boolean missingReference() {
		return true;
	}
}
```

Reply:
> Any idea why the test would fail for this case? I see that the method body of main is retained but still the class defined in the lambda is not processed such the method reference is not found.

Something with the lambda probably, method references and lambdas have a lot of bugs around them from my experience.

I'll be on vacation next week, I'll check the case above once I'm back.

Reply:
Looks like #438 to an extent. During the following stack trace, `FunctionalExpression.binding` is `null` and so the indexer "skips" the lambda:

```
"Java indexing" #61 daemon prio=5 os_prio=0 cpu=923.10ms elapsed=207.80s tid=0x00007ffff0c2e000 nid=0x4b4e runnable  [0x00007fff59ca0000]
   java.lang.Thread.State: RUNNABLE
        at org.eclipse.jdt.internal.core.search.indexing.SourceIndexer.indexResolvedDocument(SourceIndexer.java:226)
        at org.eclipse.jdt.internal.core.search.JavaSearchParticipant.indexResolvedDocument(JavaSearchParticipant.java:103)
        at org.eclipse.jdt.internal.core.search.indexing.IndexManager.indexResolvedDocument(IndexManager.java:677)
        at org.eclipse.jdt.internal.core.search.indexing.IndexManager$2.execute(IndexManager.java:1303)
        at org.eclipse.jdt.internal.core.search.processing.JobManager.run(JobManager.java:478)
        at java.lang.Thread.run(java.base@11.0.10/Thread.java:834)
```

Reply:
OK looks like the project doesn't have the JRE on its classpath, so `Runnable` is not resolved and then the rest of the problems I mentioned follow. I'll try to get the JRE on the classpath... unfortunately that is not so trivial. E.g. here I had to commit a `.classpath` file to do JRE type searches:

https://github.com/eclipse-jdt/eclipse.jdt.core/pull/333/files#diff-c6d731dbfd19fc5fe3c6c0c41bc57229e6d579e3f37f2af0507b3cb2d7b1a634

Reply:
> Looks like https://github.com/eclipse-jdt/eclipse.jdt.core/issues/438 to an extent

Curious since it's only a single file while the bug seems to rely on multiple files to be present.

Chewing on the question whether the source indexer is the correct tool to resolve the bindings ... I wonder how it can work together with `var` typed locals that are used in method references, e.g. 
```
  var someList = new ArrayList<String>();
  withAcceptor(someList.subList(0,1)::add);
```
Is the source indexer in the position to compute the correct type information for "all" shapes of local vars such that method references can work?

Reply:
> Chewing on the question whether the source indexer is the correct tool to resolve the bindings ... I wonder how it can work together with `var` typed locals that are used in method references, e.g.
> 
> ```
>   var someList = new ArrayList<String>();
>   withAcceptor(someList.subList(0,1)::add);
> ```
> 
> Is the source indexer in the position to compute the correct type information for "all" shapes of local vars such that method references can work?

I'm not sure I understand your question. You expect `SourceIndexer` to struggle with this example? There is no problem with this snippet:

```
package test;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class TestVarMethodReference {

	static class TestList<T> {

		private final List<T> list;

		TestList() {
			this(new ArrayList<>());
		}

		private TestList(List<T> list) {
			this.list = list;
		}

		public void add(T o) {
			list.add(o);
		}

		public TestList<T> sub(int s, int e) {
			return new TestList<>(list.subList(s, e));
		}

		@Override
		public String toString() {
			return list.toString();
		}
	}

	public static void main(String[] args) {
		var list = new TestList<String>();
		list.add("1");
		list.add("2");
		list.add("3");
		list.add("4");
		withAcceptor(list.sub(0, 1)::add);
		System.out.println(list);
	}

	private static void withAcceptor(Consumer<String> l) {
		l.accept(".1");
		l.accept(".2");
	}
}
```

The method reference to the `add` method is found.

I'm guessing everything is resolved here and so the method reference is also "seen".